jvm/src/main/kotlin/app/cash/trifle/Trifle.kt

```diff
@@ -15,4 +29,70 @@ object Trifle {
     EndEntityDelegate by delegate {
     constructor(keysetHandle: KeysetHandle) : this(TinkDelegate(keysetHandle))
   }
+
+  fun SignedData.verify(certAnchor: ByteArray): Boolean {
+    val TRIFLE_SIGNING_MESSAGE_VERSION: Int = 0
+
+    if (certAnchor == null || certAnchor.isEmpty()) {
+      throw Exception("Invalid input root cert")
+    }
+
+    // Validate and extract signed data
+    //   `        raw_data = raw_data,
+    //           signature = signature,
+    //           certificates = certificates,
+    val messageIn = checkNotNull(enveloped_data)
+    val signatureIn = checkNotNull(signature)
+    val certificatesIn = certificates
+
+    val signingMessage = SignedData.EnvelopedData.ADAPTER.decode(messageIn)
+
+    // Validate and extract signing message
+    //       builder.version = version
+    //       builder.signing_algorithm = signing_algorithm
+    //       builder.data_ = data_
+    check(signingMessage.version == TRIFLE_SIGNING_MESSAGE_VERSION)
+    val signingAlgorithm = when (signingMessage.signing_algorithm) {
+      SignedData.Algorithm.ECDSA_SHA256 -> TrifleAlgorithmIdentifier.ECDSASha256AlgorithmIdentifier
+      else -> {
+        throw Exception("Unsupported signing algorithm")
+      }
+    }
+
+    val cf = CertificateFactory.getInstance("X.509");
+    val root: Certificate? = cf.generateCertificate(ByteArrayInputStream(certAnchor))
+    val list = ArrayList<Certificate>()
+
+    // Set up certPath. Remove root if in certPath
+    for(i in 1..certificatesIn.size) {
+      val cert = cf.generateCertificate( ByteArrayInputStream(certificatesIn[i-1].certificate?.toByteArray() ))
+      list.add(cert)
+    }
+    list.remove(root)
+    val certPath = cf.generateCertPath(list)
```

Comment:
hmm why do we remove root when generating the cert path?

Reply:
We have to remove so certPathValidator can verify the chain using the TrustAnchor we are setting up.
If we have the trust anchor in the chain, it fails.